egg-herbie/main.rkt

```diff
@@ -215,6 +217,38 @@
 ;; egraph -> string
 (define-eggmath egraph_serialize (_fun _egraph-pointer -> _rust/datum))
 
+;; egraph -> uint
+(define-eggmath egraph_size (_fun _egraph-pointer -> _uint))
+
+;; egraph -> id -> uint
+(define-eggmath egraph_eclass_size (_fun _egraph-pointer _uint -> _uint))
+
+;; egraph -> id -> idx -> uint
+(define-eggmath egraph_enode_size (_fun _egraph-pointer _uint _uint -> _uint))
+
+;; egraph -> u32vector
+(define-eggmath
+ egraph_get_eclasses
+ (_fun [e : _egraph-pointer] [v : _u32vector = (make-u32vector (egraph_size e))] -> _void -> v))
+
+;; egraph -> id -> u32 -> (values (or symbol number) u32vector)
+(define-eggmath egraph_get_node
+                (_fun [e : _egraph-pointer]
+                      [id : _uint32]
+                      [idx : _uint32]
+                      [v : _u32vector = (make-u32vector (egraph_enode_size e id idx))]
+                      ->
+                      [f : _rust/string]
```

Comment:
So in reality these are symbols, which do have identity, so I wonder if it would be possible to directly pass in symbol pointers here. It might be as easy as declaring this as `_racket` and treating those as opaque pointers. (This might be over-optimizing.) Just a random thought.

Reply:
It can also be a number, but less conversion should be the goal.

Reply:
Yes, true. A wilder thought: the only reason we need to translate these to strings or whatever _at all_ is that we need: 1) equality, and 2) constant folding. We could pass a Racket function over to Rust to call as a callback for both. (This is what `snake_egg` does and it seems to work fine.) And then we don't need to do any conversion to fn names / numbers at all, we can directly pass pointers in both directions. We'd only need to worry about encoding the egraph data itself, which is becoming increasingly efficient.